### PR TITLE
[4.0] Decrease the amount of node saves

### PR DESCRIPTION
**Pull request by scottwulf**
Rebase of PR #80
Only save the node when attribute(s) have changed
Do not fetch monitor secret if we're the master

### CODE REVIEW

**Review by nicolasbock — DISMISSED** [review]

File: chef/cookbooks/ceph/recipes/mon.rb [hunk 1]
```diff
@@ -85,10 +87,11 @@
       add_key.run_command
       add_key.error!
 
+      # no need to check if the attribute is already set: it's part of the
```
Comment (Itxaka) [inline chef/cookbooks/ceph/recipes/mon.rb]: comment not finished

File: chef/cookbooks/ceph/recipes/osd.rb [hunk 2]
```diff
@@ -127,7 +137,7 @@
         end
         device["device"] = d.name
         node.set["ceph"]["osd_devices"].push(device)
-        node.save
+        dirty = true
```
Comment (Itxaka) [inline chef/cookbooks/ceph/recipes/osd.rb:140]: We were saving here explicitly for a reason, probably becasue we are using this values `["ceph"]["osd_devices"]` immediately below.

Are we sure this is still working as before? I would say yes, as its only used in this recipe, but would like confirmation that this has been tested properly and the devices are being picked up in the recipe afterwards without saving the node.

**Review by Itxaka — DISMISSED** [review]
Looks good, just a question and a nit

**Review by Itxaka — COMMENTED** [review]
Looks good, just a question and a nit

File: chef/cookbooks/ceph/recipes/osd.rb [hunk 3]
```diff
@@ -71,7 +71,7 @@
   if is_crowbar?
     dirty = false
 
-    node.set["ceph"]["osd_devices"] ||= []
+    node.set["ceph"]["osd_devices"] = [] if node["ceph"]["osd_devices"].nil?
```
Comment (nicolasbock) [inline chef/cookbooks/ceph/recipes/osd.rb]: Same question as is in https://github.com/crowbar/crowbar-ceph/pull/103. Why the change?

File: chef/cookbooks/ceph/recipes/osd.rb (hunk 2 above)
Comment (scottwulf) [reply chef/cookbooks/ceph/recipes/osd.rb:140]: Yes, the array of devices is enumerable (i.e. usable) immediately below even after moving node.save to the end of the file. The node.save is only needed to persist the values for subsequent executions of the cookbook.

Comment (Itxaka) [reply chef/cookbooks/ceph/recipes/osd.rb:140]: Thanks!

**Review by Itxaka — APPROVED** [review]

**Review by nicolasbock — APPROVED** [review]
